main/src/ca/uwaterloo/flix/language/phase/Namer.scala

```diff
@@ -1324,35 +1312,68 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
   }
 
   /**
-    * Performs naming on the given type parameters parameters `tparam0` under the given type environment `tenv0`.
+    * Performs naming on the given type parameters `tparam0` from the given cases `cases`.
     */
-  private def getTypeParams(tparams0: WeededAst.TypeParams, fparams0: List[WeededAst.FormalParam], tpe: WeededAst.Type, loc: SourceLocation)(implicit flix: Flix): Validation[List[NamedAst.TypeParam], NameError] = {
+  private def getTypeParamsFromCases(tparams0: WeededAst.TypeParams, cases: List[WeededAst.Case], loc: SourceLocation)(implicit flix: Flix): Validation[List[NamedAst.TypeParam], NameError] = {
     tparams0 match {
-      case WeededAst.TypeParams.Elided => getImplicitTypeParams(fparams0, tpe, loc)
-      case WeededAst.TypeParams.Explicit(tparams0) => getExplicitTypeParams(tparams0).toSuccess
+      case WeededAst.TypeParams.Elided => Nil.toSuccess // TODO allow implicit tparams?
+      case WeededAst.TypeParams.Explicit(tparams0) =>
+        mapN(getImplicitTypeParamsFromCases(cases, loc)) {
+          implicitTparams => getExplicitTypeParams(tparams0, implicitTparams)
+        }
     }
   }
 
   /**
-    * Returns the explicit type parameters from the given type parameters.
+    * Performs naming on the given type parameters `tparams0` from the given formal params `fparams` and overall type `tpe`.
     */
-  private def getExplicitTypeParams(tparams0: List[Name.Ident])(implicit flix: Flix): List[NamedAst.TypeParam] = tparams0 map {
-    case p =>
-      // Generate a fresh type variable for the type parameter.
-      // We use a kind variable since we do not know the kind of the type variable.
-      val tvar = Type.freshVar(Kind.freshVar())
-      // Remember the original textual name.
-      tvar.setText(p.name)
-      NamedAst.TypeParam(p, tvar, p.loc)
+  private def getTypeParamsFromFormalParams(tparams0: WeededAst.TypeParams, fparams: List[WeededAst.FormalParam], tpe: WeededAst.Type, loc: SourceLocation, allowElision: Boolean)(implicit flix: Flix): Validation[List[NamedAst.TypeParam], NameError] ={
```

Comment:
I am not sure where this exactly goes, but where do we account for the fact that the return type may also implicitly introduce type variables? E.g. `def f(): a = ???`.

Reply:
The parameter `tpe` is the overall type of the function E.g. `Unit -> a`. So we get the return type from that.